README.md
# Shareable Kanban Boards

[![Marketplace Version](https://badgen.net/vs-marketplace/v/iulian-radu-at.sharable-kanban-boards)](https://marketplace.visualstudio.com/items?itemName=iulian-radu-at.sharable-kanban-boards)
[![Installs](https://badgen.net/vs-marketplace/i/iulian-radu-at.sharable-kanban-boards)](https://marketplace.visualstudio.com/items?itemName=iulian-radu-at.sharable-kanban-boards)
[![Rating](https://badgen.net/vs-marketplace/rating/iulian-radu-at.sharable-kanban-boards)](https://marketplace.visualstudio.com/items?itemName=iulian-radu-at.sharable-kanban-boards)
<a href="http://opensource.org/licenses/GPL-3.0" target="_blank" rel="noreferrer noopener"><img src="https://img.shields.io/badge/license-GPL-orange.svg?color=blue&amp;style=flat-square" alt="The GPL-3.0 License"></a>

What if you can habe Kanban boards stored in the same git repo with your project?
It will be great.
Especially because in this way you no longer need to pay for and manage additional tools, and you avoid the effort to mantain the tickets in the sync with the effective progress of the project.

This extension is doing exactly this. Also it is free of charge and it will remain so.
The learning curve of using it is no curve at all, as using it is extremly intuitive.
We also added a few key features for helping you to have a very good visibility over your project.

This extension is an amazing addition for any (new) start-up and all other types of companies.
Start using it today for gaining all beneficies offered by it.

## Usage

This extension makes available the following commands:

- Shareable Kanban boards: Create a new board
- Shareable Kanban boards: Open a board
- Shareable Kanban boards: Delete a board

See below the section **"How to use a Shareable Kanban board"** for a detailed description of how to use this extension.

## Location of saved boards

All Kanban boards are saved in .vscode/sharable-kanban-boards.json and hence you can commit it to your repo.
If you do manual changes in this file, take care to do not corrupt its structure.

## Requirements

There are no special requirements.

## Extension Settings

There are not any custom settings.

## Known Issues

There are not any known issues.

## Change Log

See Change Log [here](CHANGELOG.md)

## Issues

Submit an [issue](https://github.com/iulian-radu-at/sharable-kanban-boards/issues) if you find any bug or have any request.

## Contribution

Fork the [repo](https://github.com/iulian-radu-at/sharable-kanban-boards) and submit pull requests.

---

## How to use a Shareable Kanban board

## Create a board

![Create a board](images/CreateBoard.png)

An input field asks you for the name of the new board. In case this name is already used by a board, a message will be displayed.

![Board name already exists](images/BoardNameExists.png)

If there exists not a .vscode/sharable-kanban-boards.json file, then a new one will be created which will contains a "Demo board".

## Open a board

![Open a board](images/OpenBoard.png)

A list of all available boards is presented to you so you can easily specify which one you want to open.

If there exists not a .vscode/sharable-kanban-boards.json file, then a new one will be created which will contains a "Demo board". This board will be available in the list of available boards.

## Delete a board

![Delete a board](images/DeleteBoard.png)

A list of all available boards is presented to you so you can easily specify which one you want to delete.

Attention, this operation cannot be reversed.

## Screens

There are four screens:

- Board - It displays all the tickets of a board, grouped in columns by their statuses.
- Task - It allows you to add or edit a task and to add comments.
- Columns - It allows you to manage all defined statuses and the allowed transitions of a ticket between them.
- Timeline - It displays the relations between linked tickets in a clear visual way.

## Board

![Board](images/Board.png)

The board is the core of this extension.
Here you see all the tickets of a board grouped by statuses in columns, and from it you can navigate to the Task, Columns and Timeline screens.

Use the New Task button to create a new ticket in the Task screen. Also you can access the Task screen through the use of a ticket's menu Comment or Edit command.

![Task's menu](images/CardMenu.png)

To delete a ticket use the ticket's menu.
The operation of deletion requires you to click it twice, as this operation cannot be reverted.
As such, this mecanism protects you from the accidental removal of a ticket.

![Confirm delete](images/CardMenuConfirmDelete.png)

You can filter the tasks by searching for matches in different fields of the tickets.
If no field is specified, then it is assumed to be the title field.
The results are all the tickets matching all the filters.

The following fields are recognized. Any unknown field name is considered to be part of the plain text phrase.
Between the field name and the value (which can be a single word or a phrase, if it is between double quotes) can be any number of spaces.
You cannot escape a double quote inside of a phrase.

| Field name  | Possible values or example         |
| ----------- | ---------------------------------- |
| assignee:   | name of a user                     |
| comment:    | text                               |
| dependency: | title of a ticket                  |
| done:       | today, tomorrow, 12d               |
| due:        | today, tomorrow, 3d                |
| priority:   | lowest, low, normal, high, highest |
| progress:   | ?%, 0%, 25%, 50%, 75%, 100% \*     |
| status:     | name of a column name              |
| tag:        | a tag                              |

\*Priority "?%" means a ticket without a priority set.

The following filter 'word field: "value value2" more words field2: value3 unknown_field: no_value final words' is splitted in:

- word
- field: value value2
- more words
- field2: value3
- unknown_field: no_value final words

It will match all the tickets which:

- have "value value2" as part of their "field" field
- have "value3" as part of their "field2" field
- title matches all three criterion in the same time: "word", "more words", "unknown_field: no_value final words"

![Filter tasks](images/FilteredBoard.png)

## Task

In this screen you can fill or change the fields of a ticket.
The following fields are available:

| Field name  | Meaning                                                                         |
| ----------- | ------------------------------------------------------------------------------- |
| Title       | The subject of the ticket.                                                      |
| Status      | One of the column names defined in Columns screen.                              |
| Progress    | Not defined or 0%, 25%, 50%, 100%.                                              |
| Tags        | Any number of tags.                                                             |
| Description | A more extensive text in Markdown syntax.                                       |
| Priority    | One of Lowest, Low, Normal, High or Highest.                                    |
| Due date    | The latest date when this ticket should be "done".                              |
| Assignees   | A freely defined set of user names.                                             |
| Depend on   | A set of tasks which should be "done" before this one can be considered "done". |
| Comments    | A set of comments made by the users in connection with this ticket.             |

![Task](images/Task.png)

### Progress

If the progress is 25%, 50% or 100%, an estimation done date is computed and displayed in board.
The starting date used for computing the done date is the last date when the ticket was set to have 0% progress or it was moved in a different column (status).

### Tags

A tag can be a single word or a phrase.
A set of tags can be used for example to specify ticket types.
A different set can be used for displaying the estimation of the tickets.

### Description

In this field you need to use the [Markdown](https://www.markdownguide.org/basic-syntax/) syntax. A "cheet sheet" of it can be found [here](http://markdown-it.github.io/).

### Priority

Each priority has a color, which is displayed on the side of the tickets in the board.

### Depend on

Use the screen Timeline to see the dependencies between the tickets, including when they are scheduled to be "done" or expected to be "done".

### Comments

The text in a comment is in plain text, and not Markdown.
The user name used for the author of the comment is extracted from the output of the "git config user.name" command.
If no name is found, the author is marked as "unknown".

## Columns

![Columns](images/Columns.png)

Here you can define the columns, aka statuses, and the allowed transitions of a ticket between them.

To remove a column, you will need to press twice on the delete button. This is a mechanism to prevent accidental removal of a column, as this operation cannot be reverted.

A diagram of allowed transitions helps you to have a better overwiew of them.

The allowed destinations for a dragged ticket in the board is defined by these allowed transitions and the status of the linked (depend on) tickets.
A ticket cannot be moved to a column more to the right than to the leftmost of all depending tickets' columns.

## Timeline

![Timeline](images/Timeline.png)

Here are displayed the relations between the tickets, together with their due and estimated done dates.
If a ticket has both a due and an estimated done date, only the due date will be displayed.

The status of the tickets (in which columns are they located) is represented with different colors.
A legend at the top of the page, explains the associations between colors and statuses.

---

## FAQ

### How can I create tickets of different types?

Use a tag having the type of the ticket.
For example, use the tag "Story" for defining a ticket holding a story.
You can add the tags describing a ticket's type to the tags associated to the board.
They are defined in .vscode/sharable-kanban-boards.json, field "tags" of a board.

### How can I see only tickets of a certain type?

If you are using tags to mark the type of ticket, you can filter the tasks using the "tag:".
For example, to find all tickets of type "Story" you can filter by "tag:story".

### How can I assign estimations in days to a ticket?

Use a tag representing the estimation. The tag should be a number suffixed with d (from days).
When the ticket is moved in an active status, set the due date to the corresponding duration.
When the ticket is moved in an inactive status, remove the due date.

### How can I assign estimations in story points to a ticket?

Use a tag representing the estimation.
If you use animals, then use the name of the animal as tag ("Dog").
If you are using story points, then use it prefixed with "E:" so the people known that this tag represents an estimation in story points ("E:10").
If you are using days, then use it prefixed with "E:" and suffied with "d" so the people known that this tag represents an estimation in days ("E:.5d").

### How can I define multiple sprints?

A board can contain multiple Sprints.
You can create a column for each Sprint you want to have defined.
Use these Sprint columns as they are the "To do" or the "Ready" column of a Sprint.
Order the Sprints in order from right to left.
The Sprint column from the far right represents the current sprint.

Bellow you can see how should be organized a Kanban board having one active sprint and two comming sprints.

| Backlog | Sprint 23 | Sprint 22 | Sprint 21 | In progress | Blocked | In testing | Done   |
| ------- | --------- | --------- | --------- | ----------- | ------- | ---------- | ------ |
| T-2000  | T-1900    | T-1809    | T-1745    | T-1789      |         | T-1723     | T-1777 |
| T-3000  | T-1910    | T-1807    |           |             |         |            | T-1708 |
| T-4000  | T-1905    |           |           |             |         |            |        |

### How I close a sprint?

All the tickets in "Done" and "Canceled" state must be deleted.
The column for the active sprint must be removed.
Move all tickets still in the closed Sprint column to the column of the next sprint.
Note: it must be empty when it is removed, otherwise all the tickets in it will no longer be accessible.
Create a new Sprint column on the right of Backlog column so the number of defined Sprint columns is constant.
Decide what should happen with all tickets present in the other columns.
Create the rule for allowing moving tickets from Backlog into the new created Sprint column.

| Backlog | Sprint 24 | Sprint 23 | Sprint 22 | In progress | Blocked | In testing | Done |
| ------- | --------- | --------- | --------- | ----------- | ------- | ---------- | ---- |
| T-2000  |           | T-1900    | T-1809    | T-1789      |         | T-1723     |      |
| T-3000  |           | T-1910    | T-1807    |             |         |            |      |
| T-4000  |           | T-1905    | T-1745    |             |         |            |      |

### How can I keep the not finished tickets in the next sprint?

Apply the technique described above for closing a sprint.
The tickets will be practicaly assigned to the next sprint.
Change their status if necessary.
The tickets in "Blocked" status should, eventually, be moved to a future sprint or in Backlog.
In this case they should have associated a tag called "Blocked" so their status will not be lost.
Also document as a comment why it is blocked, if you did not already did that.

### Why to use multiple boards?

You can use a board for the team and one for each member of the team.

The PO can use the board to keep track of all the epics (s)he needs to implement according to the roadmap.

The Scrum Master can use the board of the team to check the progress and what blocks the users.

Any subtask of a ticket can be present only in a "personal" board.
Once all subtasks of a ticket are done, they can be removed from the personal board and the corresponding ticket, found in the team board, moved in the corresponding state/column.


CHANGELOG.md
# Change Log

All notable changes to the "Shareable Kanban boards" extension will be documented in this file.

## 1.0.0 (2021-05-12)

### Added

- Create a new board
- Open a board
- Delete a board

### Changed

- Nothing

### Removed

- Nothing

## 1.0.1 (2021-05-14)

### Added

- Nothing

### Changed

- Indentation of list items

### Removed

- Nothing

## 1.0.2 (2021-06-02)

### Added

- Nothing

### Changed

- Fixed typo in extension's name

### Removed

- Nothing

## 1.0.3 (2021-06-02)

### Added

- Nothing

### Changed

- Fixed typo in multipe places

### Removed

- Nothing

## 1.1.0 (2021-08-30)

### Added

- Nothing

### Changed

- Changed some UI elements
- When you drag a ticket, only valid target columns remains displayed

### Removed

- Nothing

## 1.1.1 (2021-09-02)

### Added

- Nothing

### Changed

- Corrected the bug related to not loading and saving the tasks

### Removed

- Nothing

## 1.1.2 (2022-02-04)

### Added

- Nothing

### Changed

- Updated packages
- Added the License file

### Removed

- Nothing

## 1.1.3 (2022-06-20)

### Added

- Support link for extension's author (https://www.patreon.com/iulian_radu_at)

### Changed

- Fixed colors in UI for dark mode
- Updated packages

### Removed

- Nothing

## 1.1.4 (2022-09-25)

### Added

- Nothing

### Changed

- Updated packages

### Removed

- Nothing

## 1.1.5 (2023-06-10)

### Added

- Nothing

### Changed

- Updated packages
- Fixed the badges

### Removed

- Nothing


src-ext/types.ts
export type TJson = Record<string, TBoard>;

export interface TBoard {
  allowedSources: Record<string, string[]>;
  name: string;
  statuses: string[];
  tags: string[];
  tasks: TTask[];
}

enum TaskRelation {
  /* The other ticket must be started for this ticket to start */
  STARTED_TO_START,
  /* The other ticket must be started for this ticket to end */
  STARTED_TO_FINISH,
  /* The other ticket must be completed for this ticket to start */
  FINISHED_TO_START,
  /* This ticket can’t end unless the other ticket has ended */
  FINISHED_TO_FINISH,
  /* It is a subtask of the other task */
  SUBTASK_OF,
  /* It is related in an unknown mode */
  RELATES_TO,
}

export interface TTask {
  assignees?: string[];
  comments?: TComment[];
  dependOn?: string[];
  description?: string;
  dueDate?: Date;
  id: string;
  /* 1 (lowest) - 2 (low) - 3 (normal)[default] - 4 (high) - 5 (highest) */
  priority: number;
  /* 0 - 100 */
  progress?: number;
  /* it is set when we move in a new column and there is no progress set OR the progress is set to 0  */
  startDate?: Date;
  /* it is a column id */
  status: string;
  tags?: string[];
  title: string;
}

interface TComment {
  by: string;
  on: Date;
  text: string;
}


src-ext/reactPanel.ts
import * as path from 'path';
import * as vscode from 'vscode';
import { getBoard, saveBoard } from './commands';
import { getGitUserName } from './git';
import { TBoard } from './types';

export class ReactPanel {
  private static currentPanel: vscode.WebviewPanel | undefined = undefined;

  public static createOrShow(context: vscode.ExtensionContext, boardName: string): void {
    let columnToShowIn = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.viewColumn : undefined;
    if (columnToShowIn === undefined) {
      columnToShowIn = vscode.ViewColumn.One;
    }

    /* If we already have a panel, show it in the target column */
    if (this.currentPanel) {
      this.currentPanel.reveal(columnToShowIn);
      return;
    }

    /* Otherwise, create a new panel */
    this.currentPanel = vscode.window.createWebviewPanel(
      'ShareableKanbanBoard',
      'Shareable Kanban board',
      columnToShowIn,
      {
        /* Enable javascript in the webview */
        enableScripts: true,
        /* to keep the status when we switch between tabs */
        retainContextWhenHidden: true,
        /* And restrict the webview to only loading content from our extension's "out" directory */
        localResourceRoots: [vscode.Uri.file(path.join(context.extensionPath, 'out'))],
      }
    );

    const board = getBoard(boardName);
    this.currentPanel.webview.html = ReactPanel.getHtmlForWebview(context.extensionPath, board);

    this.currentPanel.webview.onDidReceiveMessage(
      (message) => {
        switch (message.command) {
          case 'save-kanban-board':
            saveBoard(message.payload);
            return;
        }
      },
      this,
      context.subscriptions
    );

    /* Reset when the current panel is closed */
    this.currentPanel.onDidDispose(() => (this.currentPanel = undefined), null, context.subscriptions);
  }

  private constructor() {}

  private static getHtmlForWebview(extensionPath: string, board: TBoard): string {
    const mainScript = 'webview.js';
    const mainStyle = 'webview.css';

    const scriptPathOnDisk = vscode.Uri.file(path.join(extensionPath, 'out', mainScript));
    const scriptUri = scriptPathOnDisk.with({ scheme: 'vscode-resource' });
    const stylePathOnDisk = vscode.Uri.file(path.join(extensionPath, 'out', mainStyle));
    const styleUri = stylePathOnDisk.with({ scheme: 'vscode-resource' });

    /* Use a nonce to whitelist which scripts can be run */
    const nonce = getNonce();
    const baseUrl = vscode.Uri.file(path.join(extensionPath, 'out')).with({ scheme: 'vscode-resource' });

    const gitUserName = getGitUserName();
    const boardJson = JSON.stringify(board, null, 2);

    const html = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no">
    <meta name="theme-color" content="#000000">
    <title>Kanban Boards</title>
    <link rel="stylesheet" type="text/css" href="${styleUri}">
    <meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src vscode-resource: https:; script-src 'nonce-${nonce}';style-src vscode-resource: 'unsafe-inline' http: https: data:;">
    <base href="${baseUrl}/">
    <script nonce="${nonce}">
      gitUserName = ${gitUserName ? `\`${gitUserName}\`` : 'undefined'};
    </script>
    <script nonce="${nonce}">
      kanbanBoard = ${boardJson};
    </script>
</head>
<body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id="root"></div>
    <script nonce="${nonce}" src="${scriptUri}"></script>
</body>
</html>`.trim();
    return html;
  }

  private static async setUserEmail(webview: vscode.Webview) {
    const outputChannel = vscode.window.createOutputChannel('Kanban boards');
    outputChannel.append('setUserEmail started');
    outputChannel.show();

    let session = await vscode.authentication.getSession('github', ['read:user', 'user:email'], {
      createIfNone: false,
    });
    if (session) {
      outputChannel.append(JSON.stringify({ github: session }, null, 2));
      outputChannel.show();

      webview.postMessage({
        command: 'authUserEmail',
        payload: {
          user: session.account,
          email: '?',
        },
      });
      return;
    }

    session = await vscode.authentication.getSession('microsoft', ['profile', 'email'], {
      createIfNone: false,
    });
    if (session) {
      outputChannel.append(JSON.stringify({ microsoft: session }, null, 2));
      outputChannel.show();

      webview.postMessage({
        command: 'authUserEmail',
        payload: {
          user: session.account,
          email: '?',
        },
      });
      return;
    }

    outputChannel.append('setUserEmail done');
    outputChannel.show();
  }
}

function getNonce() {
  let text = '';
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  for (let i = 0; i < 32; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}


src-ext/kanbanBoards.ts
import * as vscode from 'vscode';
import {
  getAllBoards,
  getConfigPath,
  getEmptyBoard,
  getExistingNames,
  getInputName,
  getSelectedName,
  saveAllBoards,
  saveBoard,
} from './commands';
import { ReactPanel } from './reactPanel';

export function createBoard(context: vscode.ExtensionContext) {
  createKanbanBoard(context);
}

export function openBoard(context: vscode.ExtensionContext) {
  openKanbanBoard(context);
}

export function deleteBoard() {
  deleteKanbanBoard();
}

async function createKanbanBoard(context: vscode.ExtensionContext) {
  const allBoards = getAllBoards();
  if (allBoards === undefined) {
    return;
  }

  const existingNames: string[] = getExistingNames(allBoards);

  const name = await getInputName(existingNames);
  if (name === undefined) {
    return;
  }

  saveBoard(getEmptyBoard(name));
  ReactPanel.createOrShow(context, name);
}

async function openKanbanBoard(context: vscode.ExtensionContext) {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  const allBoards = getAllBoards();
  if (allBoards === undefined) {
    return;
  }

  const existingNames: string[] = getExistingNames(allBoards);
  if (existingNames.length === 0) {
    vscode.window.showInformationMessage(`Shareable Kanban boards: no saved boards exist in "${configPath}".`);
    return;
  }

  const name = await getSelectedName(existingNames);
  if (name === undefined) {
    return;
  }

  ReactPanel.createOrShow(context, name);
}

async function deleteKanbanBoard() {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  const allBoards = getAllBoards();
  if (allBoards === undefined) {
    return;
  }

  const existingNames: string[] = getExistingNames(allBoards);
  if (existingNames.length === 0) {
    return;
  }

  const name = await getSelectedName(existingNames);
  if (name === undefined) {
    return;
  }

  const boards = getAllBoards();
  if (name in boards === false) {
    return;
  }
  delete boards[name];
  saveAllBoards(boards);
  vscode.window.showInformationMessage(`Shareable Kanban boards: Removed board '${name}'`);
}


src-ext/commands.ts
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { TBoard, TJson, TTask } from './types';

const emptyBoard: TBoard = {
  name: '',
  statuses: ['Backlog', 'To do', 'In progress', 'Review', 'Done'],
  allowedSources: {
    Backlog: ['To do'],
    'To do': ['Backlog', 'In progress'],
    'In progress': ['Backlog', 'To do', 'Review'],
    Review: ['In progress'],
    Done: ['Review'],
  },
  tags: [],
  tasks: [],
};

const demoTags = ['Epic', 'Story', 'Task', 'Subtask', 'Bug', 'Blocked'];

const demoTasks: TTask[] = [
  {
    id: 'f037e9b4ec0d417f9c69dc8929f69700',
    title: 'Shareable Kanban board',
    status: 'In progress',
    tags: ['Story'],
    priority: 3,
    description: 'Using shareable Kanban board',
    assignees: ['Me', 'Team'],
    dependOn: [
      'f037e9b4ec0d417f9c69dc8929f69701',
      'f037e9b4ec0d417f9c69dc8929f69703',
      'f037e9b4ec0d417f9c69dc8929f69704',
    ],
  },
  {
    id: 'f037e9b4ec0d417f9c69dc8929f69701',
    title: 'Install the "Shareable Kanban board" extension',
    status: 'Done',
    tags: ['Task'],
    priority: 5,
    description: 'Shareable Kanban board is available inside of Visual Studio Code',
    assignees: ['Me'],
    comments: [
      {
        by: 'Me',
        on: new Date('2021-05-05'),
        text: 'An amazing product',
      },
    ],
  },
  {
    id: 'f037e9b4ec0d417f9c69dc8929f69702',
    title: 'Learn advanced use of the shareable Kanban board',
    status: 'Backlog',
    tags: ['Task'],
    priority: 2,
    description: 'At the end I will know how to use efficiently this board',
    assignees: ['Me'],
  },
  {
    id: 'f037e9b4ec0d417f9c69dc8929f69703',
    title: 'Learn using the shareable Kanban board',
    status: 'In progress',
    tags: ['Task'],
    priority: 3,
    description: 'At the end I will know how to use efficiently this board',
    assignees: ['Me', 'Team'],
  },
  {
    id: 'f037e9b4ec0d417f9c69dc8929f69704',
    title: 'Learn to use the filtering of tasks in shareable Kanban board',
    status: 'To do',
    tags: ['Task'],
    priority: 4,
    description:
      'At the end I will know how to find tickets according to different criterion:\n* assignee:\n  * assignee: partial user name\n* comment:\n  * comment: partial phrase\n* dependency:\n  * dependency: partial ticket name\n* done:\n  * done: today\n  * done: 3d\n* due:\n  * due: tomorrow\n  * due: 10d\n* priority:\n  * priority: priority_name\n* progress:\n  * progress: 25%\n  * progress: ?%\n* status:\n  * status: status name\n* tag:\n  * tag: partial tag name\n\nAlso that each criterion makes the selection even finer.',
    assignees: ['Me', 'Team'],
  },
];

const emptyJson: TJson = {
  'Demo board': {
    ...emptyBoard,
    name: 'Demo board',
    tags: demoTags,
    tasks: demoTasks,
  },
};

export function getAllBoards(): TJson | undefined {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  try {
    const content = fs.readFileSync(configPath, 'utf8');
    const boards: TJson = JSON.parse(content);
    return boards;
  } catch (err) {
    fs.writeFileSync(configPath, JSON.stringify(emptyJson, null, 2));
    return emptyJson;
  }
}

export function getBoard(name: string): TBoard | undefined {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  try {
    const boards = getAllBoards();
    const board: TBoard | undefined = boards?.[name];
    if (board !== undefined) {
      return board;
    }

    return getEmptyBoard(name);
  } catch (e) {
    return getEmptyBoard(name);
  }
}

export function getEmptyBoard(name: string): TBoard {
  return { ...emptyBoard, name };
}

export function saveBoard(board: TBoard) {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  const allBoards = getAllBoards();
  if (allBoards === undefined) {
    return;
  }

  const newAllBoards = { ...allBoards, [board.name]: board };
  saveAllBoards(newAllBoards);
}

export function saveAllBoards(boards: TJson) {
  const configPath = getConfigPath();
  if (configPath === undefined) {
    return;
  }

  fs.writeFileSync(configPath, JSON.stringify(boards, replacer, 2));
}

function replacer(key: string, value: any): any {
  switch (key) {
    case 'dueDate':
    case 'on':
    case 'startDate':
      return typeof value === 'string' ? value : formatDate(value);
    default:
      return value;
  }
}

function formatDate(date: Date): string {
  /* yyyy-MM-dd */
  const year = date.getFullYear();
  const month = date.getFullYear();
  const day = date.getFullYear();
  return `${year}-${to2digits(month)}-${to2digits(day)}`;
}

function to2digits(value: number): string {
  return value.toLocaleString('en-US', {
    minimumIntegerDigits: 2,
    useGrouping: false,
  });
}

export function getConfigPath(): string | undefined {
  const workspaceFolders = vscode.workspace.workspaceFolders;
  if (workspaceFolders === undefined || workspaceFolders.length === 0) {
    return;
  }

  const rootPath = workspaceFolders[0].uri.fsPath;
  const vscodePath = `${rootPath}${path.sep}.vscode`;
  if (fs.existsSync(vscodePath) === false) {
    fs.mkdirSync(vscodePath);
  }
  return `${vscodePath}${path.sep}shareable-kanban-boards.json`;
}

export function getExistingNames(allBoards: TJson) {
  return Object.keys(allBoards);
}

export async function getInputName(existingNames: string[]) {
  const name = await vscode.window.showInputBox({
    placeHolder: 'Name of this new Kanban board',
    validateInput: (value: string) => getValidateInput(value, existingNames),
  });
  return name;
}

function getValidateInput(
  value: string,
  existingNames: string[]
): string | undefined | null | Thenable<string | undefined | null> {
  return existingNames.includes(value) ? 'This name is already used' : null;
}

export async function getSelectedName(existingNames: string[]) {
  const name = await vscode.window.showQuickPick(existingNames, {
    placeHolder: 'Name of an existing Kanban board',
    canPickMany: false,
  });
  return name;
}
